designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,627 @@
+* `parser` - an object in `eslint.config.js` files (a string in `.eslintrc`)
+* `processor` - an object in `eslint.config.js` files (a string in `.eslintrc`)
+
+Each of these keys used to require one or more strings specifying module(s) to load in `.eslintrc`. In `eslint.config.js`, these are all objects, requiring users to manually specify the objects to use.
+
+The following keys are invalid in `eslint.config.js`:
+
+* `env` - responsibility of the user
+* `overrides` - responsibility of the user
+* `plugins` - replaced by `ruledefs`
+* `root` - always considered `true`
+
+Each of these keys represent different ways of augmenting how configuration is calculated and all of that responsibility now falls on the user.
+
+#### Referencing Plugin Rules
+
+The `plugins` key in `.eslintrc` was an array of strings indicating the plugins to load, allowing you to specify processors, rules, etc., by referencing the name of the plugin. It's no longer necessary to indicate the plugins to load because that is done directly in the `eslint.config.js` file. For example, consider this `.eslintrc` file:
+
+```yaml
+plugins:
+  - react
+
+rules:
+  react/jsx-uses-react: error
+```
+
+This file tells ESLint to load `eslint-plugin-react` and then configure a rule from that plugin. The `react/` is automatically preprended to the rule by ESLint for easy reference.
+
+In `eslint.config.js`, the same configuration is achieved using a `ruledefs` key:
+
+```js
+const reactPlugin = require("eslint-plugin-react");
+
+exports.config = {
+    files: ["*.js"],
+    ruledefs: {
+        react: reactPlugin.rules
+    },
+    rules: {
+        "react/jsx-uses-react": "error"
+    }
+};
+```
+
+Here, it is the `ruledefs` that assigns the name `react` to the rules from `eslint-plugin-react`. The reference to `react/` in a rule will always look up that value in the `ruledefs` key.
+
+#### Referencing Parsers and Processors
+
+In `.eslintrc`, the `parser` and `processor` keys required strings to be specified, such as:
+
+```yaml
+plugins: ["markdown"]
+parser: "babel-eslint"
+processor: "markdown/markdown"
+```
+
+In `eslint.config.js`, you would need to pass the references directly, such as:
+
+```js
+exports.config = {
+    files: ["*.js"],
+    parser: require("babel-eslint"),
+    processor: require("eslint-plugin-markdown").processors.markdown
+};
+```
+
+In both cases, users now must pass a direct object reference. This has the benefit of using the builtin Node.js module resolution system or allowing users to specify their own. There is never a question of where the modules will be resolved from.
+
+#### Applying an Environment
+
+Unlike with `.eslintrc` files, there is no `env` key in `eslint.config.js`. Users can mimic the behavior of `env` by assigning directly to the `globals` key:
+
+```js
+const globals = require("globals");
+
+exports.config = {
+    files: ["*.js"],
+    globals: {
+        MyGlobal: true,
+        ...globals.browser
+    }
+};
```

Comment:
even if this is a separate package, that'd be great

Reply:
I'm not completely against the idea of including those globals in ESLint directly, but I'm not sure it has a ton of added value over just including the `globals` package. Ultimately, we'd just end up pulling in all of this from `globals` anyway rather than managing it ourselves. I'd like to hear some more opinions on this.

(Also, this is one of those things we can always add later if it turns out to be a pain point -- I don't know that it needs to be part of this design right now.)

Reply:
This is resolved by https://github.com/eslint/rfcs/pull/9/files#diff-4fab19e15273179c525fc49a689b73a5R447, though I think it's important to note that this will only work for the default parser.

Reply:
This really isn't related parsers. Parsers don't know or care about globals, it's only ESLint that cares. By moving the `ecmaVersion` to a language-level property, ESLint (not the parser) will be in charge of defining globals so even custom parsers will benefit.